R8127 ethernet driver

abhsahu opened the pull request:
Realtek R8127 driver can be downloaded from https://www.realtek.com/Download/List?cate_id=584

Where it is maintained as out of tree module.

This patch series adds this out of tree module in drivers/net/ethernet/realtek/r8127 and update makefile and Kconfig to get R8127  driver build from kernel build system. 

Once Realtek upstream these changes, then these patches can be reverted.

clsotog commented:
Do you have dmesg when loading this driver at the system? 

Also last line of drivers/net/ethernet/realtek/r8127/Kconfig is missing "r". Driver module will be called r8127.

nvmochs commented:
@abhsahu Couple of questions...

- What are the upstream plans for this driver?

- This series is still leaving a lot of OOT driver support in place, e.g. there are a lot of ifdefs to handle different Linux kernel versions and features within the driver. Is the thought that this is carried until the driver is upstream and this series is reverted?

- In "NVIDIA: SAUCE: Add r8127 in kernel build”;
  - Nit: The newly added Kconfig refers to this as a “fast ethernet” adapter. Isn’t this a 10GbE NIC? 
  - How was it decided on which features to leave enabled? Was it just based on the “= y” from the OOT Makefile? That thought process should be captured in the commit message. 
 - In "NVIDIA: SAUCE: r8127: remove unused files”
   - How was it determined that these files are not required? The commit message should include more details as to what is being removed and why.

abhsahu commented:
> Do you have dmesg when loading this driver at the system?
> 

I will get the system again and share the dmesg logs.

> Also last line of drivers/net/ethernet/realtek/r8127/Kconfig is missing "r". Driver module will be called r8127.

I have fixed this and updated.

abhsahu commented:
> @abhsahu Couple of questions...
> 
> * What are the upstream plans for this driver?

I checked with Realtek and their tentative timeline to start the upstreaming process in next month (May 2025).

> * This series is still leaving a lot of OOT driver support in place, e.g. there are a lot of ifdefs to handle different Linux kernel versions and features within the driver. Is the thought that this is carried until the driver is upstream and this series is reverted?

Given that we plan to revert this series once the driver is upstreamed, we have planned to leave those ifdefs. It will leave us with unused code for sometime but it should not cause any issue. Correct ? 

> * In "NVIDIA: SAUCE: Add r8127 in kernel build”;
>   
>   * Nit: The newly added Kconfig refers to this as a “fast ethernet” adapter. Isn’t this a 10GbE NIC?

I have updated the Kconfig description and fixed that.

>   * How was it decided on which features to leave enabled? Was it just based on the “= y” from the OOT Makefile? That thought process should be captured in the commit message.

Yes. It is based on "=y" from OOT makefile. I have update the commit message now and mentioned it.
 
> * In "NVIDIA: SAUCE: r8127: remove unused files”
>   
>   * How was it determined that these files are not required? The commit message should include more details as to what is being removed and why.

I have squashed this commit into previous code where all other flags are being removed.
These files build are dependent upon these removed flags, so we can remove the unused files also in the same commit.

abhsahu commented:
Thanks @clsotog and @nvmochs for your comments.
I have addressed that and updated the PR.

nvmochs commented:
> Given that we plan to revert this series once the driver is upstreamed, we have planned to leave those ifdefs. It will leave us with unused code for sometime but it should not cause any issue. Correct ?

Correct, it should not cause issues and is more of an annoyance, i.e. when grepping through the code. Hopefully the driver is upstreamed quickly and we can move to that.

<hr>

Thanks for the other updates, no further questions from me.

`Acked-by: Matthew R. Ochs <mochs@nvidia.com>`

abhsahu commented:
The following are the dmesg logs after loading this module for r8127.

[    5.208893] r8127 0000:01:00.0: Adding to iommu group 16
[    5.209071] r8127 Ethernet controller driver 11.014.00-NAPI loaded
[    5.209119] r8127 0000:01:00.0: enabling device (0000 -> 0003)
[    5.223622] r8127 0000:01:00.0 (unnamed net_device) (uninitialized): Invalid ether addr 00:00:00:00:00:00
[    5.223635] r8127 0000:01:00.0 (unnamed net_device) (uninitialized): Random ether addr 8e:8b:ad:ab:cd:c5
[    5.224067] r8127: This product is covered by one or more of the following patents: US6,570,884, US6,115,776, and US6,327,625.
[    5.224131] r8127  Copyright (C) 2025 Realtek NIC software team <nicfae@realtek.com>
                This program comes with ABSOLUTELY NO WARRANTY; for details, please see <http://www.gnu.org/licenses/>.
                This is free software, and you are welcome to redistribute it under certain conditions; see <http://www.gnu.org/licenses/>.
[    5.261720] r8127 0000:01:00.0 enp1s0: renamed from eth1
[   62.830328] r8127: enp1s0: link up


lspci output:

0000:01:00.0 Ethernet controller: Realtek Semiconductor Co., Ltd. Device 8127 (rev 05)
        Subsystem: Realtek Semiconductor Co., Ltd. Device 8127
        Physical Slot: 0
        Flags: bus master, fast devsel, latency 0, IRQ 161, IOMMU group 16
        I/O ports at 1000 [size=256]
        Memory at 67100000 (64-bit, non-prefetchable) [size=256K]
        Memory at 67140000 (64-bit, non-prefetchable) [size=16K]
        Capabilities: [40] Power Management version 3
        Capabilities: [50] MSI: Enable- Count=1/1 Maskable- 64bit+
        Capabilities: [70] Express Endpoint, MSI 01
        Capabilities: [b0] MSI-X: Enable+ Count=64 Masked-
        Capabilities: [d0] Vital Product Data
        Capabilities: [100] Advanced Error Reporting
        Capabilities: [148] Virtual Channel
        Capabilities: [164] Device Serial Number 00-00-00-00-00-00-00-00
        Capabilities: [174] Secondary PCI Express
        Capabilities: [184] Physical Layer 16.0 GT/s <?>
        Capabilities: [1a8] Lane Margining at the Receiver <?>
        Capabilities: [244] Latency Tolerance Reporting
        Capabilities: [24c] L1 PM Substates
        Capabilities: [25c] Data Link Feature <?>
        Capabilities: [268] Precision Time Measurement
        Capabilities: [274] Vendor Specific Information: ID=0003 Rev=1 Len=054 <?>
        Kernel driver in use: r8127
        Kernel modules: r8127

ifconfig output:

# ifconfig enp1s0

enp1s0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.16.179.156  netmask 255.255.252.0  broadcast 172.16.179.255
        inet6 fe80::453b:e7f8:9f41:6a66  prefixlen 64  scopeid 0x20<link>
        ether 8a:5c:32:58:3f:77  txqueuelen 1000  (Ethernet)
        RX packets 2620486  bytes 2789836571 (2.7 GB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1388  bytes 162044 (162.0 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device interrupt 161

clsotog commented:
These cards does not have Mac address? 
Everytime you boot you have different Mac address? 
Does the vendor will provide a way to get permanent Mac Address?

clsotog commented:
Acked-by: Carol L Soto (csoto@nvidia.com)

My last question can be answered in an email. its more a question of the final product.

nvmochs commented:
Pulled and present in https://git.launchpad.net/~canonical-kernel/ubuntu/+source/linux-nvidia/+git/noble/log/?h=nvidia-6.11-next

Closing PR.
